### PR TITLE
Add primary person list collector to schema validation and tests

### PR DESCRIPTION
**Pull request by bambrp**
### Context
As part of the effort to allow for designating a list item as a "primary person", schema changes were introduced that required validating.

This does that, and performs a series of similar tests to those performed on normal list collectors.

### How to test
Clone and run as per v3 branch code, then feed compliant and non-compliant schemas and observe the error messages or lack thereof.

### CODE REVIEW

File: app/validation/validator.py [hunk 1]
```diff
@@ -1174,7 +1261,13 @@ def _get_answers_with_parent_ids(self, json_to_validate):
     def _get_questions_with_context(self, json):
         for section in json.get('sections'):
             for group in section.get('groups'):
-                for block in group.get('blocks'):
+                if group is None:
+                    print(section)
```
Comment (bitdivision) [inline app/validation/validator.py]: As discussed, probably worth adding some more information here and returning an error, or using the logging library to allow setting level etc.

Comment (bambrp) [reply app/validation/validator.py]: I'm removing it from here and pushing it up into being called from much earlier parts of validation as it is not question-specific, and later validation methods break anyway when such basic problems exist in the schema.

File: app/validation/validator.py [hunk 2]
```diff
@@ -33,39 +33,58 @@ def validate_schema(self, json_to_validate):
...
-                validation_errors.extend(
-                    self._validate_blocks(json_to_validate,
-                                          section,
-                                          group,
-                                          all_groups,
-                                          answers_with_parent_ids,
-                                          numeric_answer_ranges)
-                )
+        try:
+            all_groups = self._validate_core_structure(json_to_validate)
```
Comment (JamesGardiner) [inline app/validation/validator.py]: This line fails flake8

File: app/validation/validator.py [hunk 3]
```diff
@@ -101,14 +118,23 @@ def _validate_blocks(self, json_to_validate, section, group, all_groups, answers
...
+                except KeyError as e:
+                    errors.append(f'Missing key in list collector: {e}')
             elif block['type'] == 'ListCollector':
                 try:
                     errors.extend(self._validate_list_collector(block))
                 except KeyError as e:
                     errors.append(f'Missing key in list collector: {e}')
-            elif block['type'] in ['ListAddQuestion', 'ListEditQuestion', 'ListRemoveQuestion']:
-                errors.append(f'Block type: {block["type"]} not allowed outside of ListCollectors')
+            elif block['type'] in ['PrimaryPersonListAddOrEditQuestion']:
```
[the remark this answers is not in the document]
Comment (bambrp) [reply app/validation/validator.py]: done

File: app/validation/validator.py (hunk 2 above)
Comment (bambrp) [reply app/validation/validator.py]: fixed

File: app/validation/validator.py [hunk 4]
```diff
@@ -377,6 +403,20 @@ def _validate_list_answer_references(self, block):
 
         return errors
 
+    def _validate_primary_person_list_answer_references(self, block):
```
[the remark this answers is not in the document]
Comment (bambrp) [reply app/validation/validator.py]: Some of this is making things less consistent with the closely matching code nearby and throughout the file, but Ive put it in.

File: app/validation/validator.py [hunk 5]
```diff
@@ -33,39 +33,58 @@ def validate_schema(self, json_to_validate):
...
         all_errors['validation_errors'] = validation_errors
 
         return all_errors
 
+    def _validate_core_structure(self, json_to_validate):
+        all_groups = []
+        all_sections = json_to_validate.get('sections', [])
+        if all_sections:
+            for section in all_sections:
+                section_groups = section.get('groups')
```
[the remark this answers is not in the document]
Comment (bambrp) [reply app/validation/validator.py]: Done

[the remark this answers is not in the document]
Comment (MebinAbraham) [reply app/validation/validator.py]: I might have missed something, but surely JSON schema should already be doing this?

File: schemas/blocks/primary_person_list_collector.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      "add_or_edit_answer": {
+        "$ref": "list_collector.json#/answer_reference"
+      },
+      "question": {
+        "$ref": "../questions/definitions.json#/question"
+      },
+      "question_variants": {
+        "$ref": "definitions.json#/question_variants"
+      },
+      "add_or_edit_block": {
```
Comment (MebinAbraham) [inline schemas/blocks/primary_person_list_collector.json]: How come this is merged into one now?. Can list collectors also be merged in future?

File: app/validation/validator.py [hunk 7]
```diff
@@ -817,8 +928,11 @@ def _validate_schema_contains_submission_page(self, schema, last_block):
         return []
 
     @staticmethod
-    def _error_message(message):
-        return {'message': 'Schema Integrity Error. {}'.format(message)}
+    def _error_message(message, ref=None):
+        error = {'message': f'Schema Integrity Error. {message}'}
+        if ref is not None and isinstance(ref, str):
```
Comment (MebinAbraham) [inline app/validation/validator.py]: The first could be removed.
```suggestion
        if isinstance(ref, str):
```

**Comment by bitdivision** [comment]
As discussed @MebinAbraham , I don't think multiple primary person list collectors will actually break anything, although the use cases are difficult to think of.

File: app/validation/validator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bambrp) [reply app/validation/validator.py]: My main motivation for the initial version was to provide better feedback on where things failed (by id or at least my showing the failing group/block's parent), make for quicker schema fixing.
Perhaps the JSON schema checker could be told to be more verbose, or swapped out for something better, which I believe has been discussed.

File: app/validation/validator.py [hunk 8]
```diff
@@ -33,39 +33,58 @@ def validate_schema(self, json_to_validate):
...
+
+            for section in json_to_validate['sections']:
+                for group in section['groups']:
+
+                    validation_errors.extend(self._validate_routing_rules(group, all_groups, answers_with_parent_ids))
+
+                    for skip_condition in group.get('skip_conditions', []):
+                        validation_errors.extend(self._validate_skip_condition(skip_condition, answers_with_parent_ids, group))
+
+                    validation_errors.extend(
```
[the remark this answers is not in the document]
Comment (bitdivision) [reply app/validation/validator.py]: We should be able move the except to immediately after the call to self._build_groups_list?

File: app/validation/validator.py (hunk 7 above)
Comment (bambrp) [reply app/validation/validator.py]: okay

File: app/validation/validator.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bambrp) [reply app/validation/validator.py]: done

**Pull request merged**